### PR TITLE
Added counties/states for EU countries and GB.

### LINKED ISSUE #24354: Unable to retrieve tax rate row when buyer's country has a county with a space in it
**Describe the bug**

Certain countries (e.g. GB) have counties which are treated as states for the purposes of things like tax rates.

If you attempt to create an entry in the tax table for such a county, and that county includes spaces in its name (e.g. Isle of Wight), after saving the tax rate table, the state code loses its spaces (e.g. becomes ISLEOFWIGHT)

Unfortunately, this means that buyers that enter "Isle of Wight" for their County during checkout will NOT see the tax rate you have set up for them in the tax table. (Not surprisingly, if they enter ISLEOFWIGHT in the checkout form, that works.)

**Screenshots**

Screen Shot 2019-08-07 at 4 42 32 PM

**Expected behavior**

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**

WooCommerce 3.7 RC1

### PR DESCRIPTION
**Pull request by peterfabian**
This should fix the problematic behaviour of automated taxes when using WooCommerce Services, Jetpack and TaxJar + WooCommerce.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### TODO:
* [ ] check that matching of shipping zones still works correctly, I saw some strange effects where it didn't match any shipping zone, even the catch-all.

### Changes proposed in this Pull Request:

Closes #24354 .
Counties and subdivisions coming from CLDR release v36: http://unicode.org/Public/cldr/36/
Where available, I used the local state/county names. 

### How to test the changes in this Pull Request:

1. Install WooCommerce with WooCommerce Services, connect Jetpack and set automated tax calculations.
2. Try to check out a couple of times using different County/State, e.g. for Country GB, try empty, multiple word and one word options.
3. Check that tax is only applied if the County/State does not have a space in it.
4. Apply branch
5. Test checkouts again, it should work fine in all cases. A dropdown with County selection should appear instead of a text input field, but it should **not** be compulsory to fill in any of the options.
6. Verify that the automatically created tax rate rules are no longer duplicated.
7. Remove all the automatically created tax rates, add manual ones and verify the checkout calculates the tax correctly.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Added counties/states for EU countries and GB to fix edge case with automated tax calculation.

### CODE REVIEW

**Comment by harriswong** [comment]
Tested against WooCommerce Services' automated tax: https://github.com/Automattic/woocommerce-services/issues/1600#issuecomment-579528796 👍

File: i18n/states.php [hunk 1]
```diff
@@ -347,8 +434,47 @@
...
+		'OCC' => __( 'Occitanie', 'woocommerce' ),
+		'PAC' => __( "Provence-Alpes-Côte d'Azur", 'woocommerce' ),
+		'PDL' => __( 'Pays de la Loire', 'woocommerce' ),
+	),
+	'GB' => array( // British countries/provinces.
+		'ENG' => __( 'England', 'woocommerce' ),
+		'NIR' => __( 'Northern Ireland', 'woocommerce' ),
+		'SCT' => __( 'Scotland', 'woocommerce' ),
+		'WLS' => __( 'Wales', 'woocommerce' ),
+	),
```
Comment (harriswong) [inline i18n/states.php]: One question about the county and these UK addresses:
```
2 Chestnut Hall Avenue, Maghaberry, Moira, County Armagh, BT67 0GG
40 Whitehall Way, Rockingham, Rotherham, South yorkshire, S61 4HW
19 Carisbrooke Close, EASTBOURNE, East Sussex, BN23 8EQ
```
I don't see "England", "Northern Ireland", "Scotland" or "Wales" in these addresses, do we assume the user to know which to pick? 

For example, using address `2 Chestnut Hall Avenue, Maghaberry, Moira, County Armagh, BT67 0GG` in this form, 
image
I am assuming we drop the `County Armagh` entry and use `Moira` as city. Or should `County Armagh` be the county?

Comment (harriswong) [reply i18n/states.php]: Also, user made a comment on the values in `county`. https://github.com/Automattic/woocommerce-services/issues/1600#issuecomment-579532076.

Since we don't have "England", "Northern Ireland", "Scotland" or "Wales" as separate countries, any of them would be selected as "United Kingdom".  I do agree that putting these 4 countries as "county" is a bit awkward.

Comment (peterfabian) [reply i18n/states.php]: There are multiple levels of subdivisions in some countries and UK is one of them. We can instead include the longer & more detailed list if that feels more natural to the UK folks, please see https://github.com/woocommerce/woocommerce/pull/25509/commits/e981529d5c16483d3910f700e16e2a369b6a6ef9

Comment (mikejolley) [reply i18n/states.php]: It's very uncommon (at least in my experience spending way too much online) to select counties in the UK like this. Usually it's just a text field. In practice, it doesn't affect postal addresses much at all, since county can be derived from the postcode (postcode being the most important field for postage). To me using a select here feels unnatural.

Surprised to hear this is a problem related to taxes; there are no per-county tax rates. Everyone in the UK pays the same rate of sales tax.

Comment (peterfabian) [reply i18n/states.php]: Thanks for chiming in, @mikejolley!

Background: WC Services automated taxes solution looks up tax rates based on country + state + city + zip code and since the state gets sanitized each time, it fails to find a match in WC, thus creating a new tax rule each time. In addition, the tax fails to be applied if there's a space in the County name.
This could be fixed in the WooCommerce services, however, I see they have their own custom db code to handle this case, which feels not right, so I thought we should provide a method in core to do the matching and they should then start using it. 

While it's true that in the UK everyone pays the same sales tax (although someone [mentioned it might not be true all the time](https://github.com/woocommerce/woocommerce/issues/24354#issuecomment-577882153)), I found out that even the same country+state+city+zip code combination can have multiple tax rates in some coutries, so I wasn't convinced that leaving out State when creating tax rules would be the best solution. Maybe it could work for the UK, but not in general...

One possibility I can see is to use some conditionals in WCS where for the UK the tax rules would only fill in country + zip code, while for other countries it could fill in other fields. 

TL;DR: Due to the nature of the bug, I'm not sure if there's an easier and better solution that would work in all cases, but I'm open to suggestions/finding a better solution.

Comment (mikejolley) [reply i18n/states.php]: Yeah @peterfabian that is tricky. My concern is how this affects customers (checkout) as the experience will be strange and could affect conversions (we'd have no way to track that though?).

Ideally the tax rate service would return wildcards, effectively indicating that city/state is not important for the tax rate and negating the matching issue. But I don't know enough about the service so..

Comment (peterfabian) [reply i18n/states.php]: As the field is optional by default, I would hope it's not going to affect the conversions.

Though another worry of mine is the huge blob of text with the list of Counties sent over at the point of checkout... it's 60kb now :/

**Comment by peterfabian** [comment]
@harriswong We discussed this some more in Proton and thought it could be better to include this changer in the WooCommerce Services plugin, as that's where the problem primarily appears in combination with automated taxes. This also means we wouldn't be unnecessarily increasing the download size of checkout for all WC customers and it would make it easier to remove the list once we implement the proper solution. What do you think?

**Comment by harriswong** [comment]
Thanks for the update @peterfabian. When our team was discussing this issue, we thought that may be WCS can nullify the county field before it sends the request over to woocommerce core's tax calculation methods. I haven't tried this solution but I think WCS can do that until the API option is available. 

I will bring this up in the Extendables.

**Comment by peterfabian** [comment]
Closing as this will be solved by the extension, which is a better way than to make WC slower for everyone.
